x/logic/predicate/json.go

```diff
@@ -32,6 +33,79 @@
 	AtomValidJSONNumber = engine.NewAtom("json_number")
 )
 
+var (
+	errWrongStreamType = errors.New("wrong stream type")
+	errWrongIOMode     = errors.New("wrong i/o mode")
+	errPastEndOfStream = errors.New("past end of stream")
+	errInvalidUTF8     = errors.New("invalid UTF-8")
+)
+
+var (
+	operationInput  = engine.NewAtom("input")
+	operationOutput = engine.NewAtom("output")
+)
+
+var (
+	permissionTypeStream          = engine.NewAtom("stream")
+	permissionTypeTextStream      = engine.NewAtom("text_stream")
+	permissionTypePastEndOfStream = engine.NewAtom("past_end_of_stream")
+)
+
+// JSONRead is a predicate that reads a JSON from a stream and unifies it with a Prolog term.
+//
+// See json_prolog/2 for the canonical representation of the JSON term.
+//
+// The signature is as follows:
+//
+//	json_read(+Stream, ?Term) is det
+//
+// Where:
+//   - Stream is the input stream from which the JSON is read.
+//   - Term is the Prolog term that represents the JSON structure.
+func JSONRead(vm *engine.VM, stream, term engine.Term, cont engine.Cont, env *engine.Env) *engine.Promise {
+	is, err := prolog.AssertStream(stream, env)
+	if err != nil {
+		return engine.Error(err)
+	}
+
+	decoder := newTextStreamDecoder(is)
+	decoded, err := decodeJSONToTerm(decoder, env)
+	if err != nil {
+		return engine.Error(err)
+	}
+	if _, err := decoder.Token(); !errors.Is(err, io.EOF) {
+		return engine.Error(
+			engine.SyntaxError(AtomSyntaxErrorJSON.Apply(AtomMalformedJSON.Apply(engine.Integer(decoder.InputOffset()))), env))
+	}
+
+	return engine.Unify(vm, term, decoded, cont, env)
+}
+
+// JSONWrite is a predicate that writes a Prolog term as a JSON to a stream.
+//
+// The JSON object is of the same format as produced by json_read/2.
+//
+// The signature is as follows:
+//
+//	json_write(+Stream, +Term) is det
+//
+// Where:
+//   - Stream is the output stream to which the JSON is written.
+//   - Term is the Prolog term that represents the JSON structure.
+func JSONWrite(_ *engine.VM, stream, term engine.Term, cont engine.Cont, env *engine.Env) *engine.Promise {
+	os, err := prolog.AssertStream(stream, env)
+	if err != nil {
+		return engine.Error(err)
+	}
+
+	buf := newTextStreamWriter(os)
+	if err := encodeTermToJSON(term, buf, env); err != nil {
+		return engine.Error(err)
+	}
+
+	return cont(env)
+}
```

Comment:
No, it’s not necessary because the correct error will be raised when accessing the stream (cf. unit tests). That said, I understand the defensive approach you suggest—checking preconditions early. Unfortunately, I can’t do that, as I don’t have a simple way to verify the stream’s properties, like its `mode` or `type` (these are private in the library)